lib/resourcemerge/admissionregistration.go

```diff
@@ -0,0 +1,72 @@
+package resourcemerge
+
+import (
+	admissionregv1 "k8s.io/api/admissionregistration/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+)
+
+// EnsureValidatingWebhookConfiguration ensures that the existing matches the required.
+// modified is set to true when existing had to be updated with required.
+func EnsureValidatingWebhookConfiguration(modified *bool, existing *admissionregv1.ValidatingWebhookConfiguration, required admissionregv1.ValidatingWebhookConfiguration) {
+	EnsureObjectMeta(modified, &existing.ObjectMeta, required.ObjectMeta)
+	ensureValidatingWebhookConfigurationDefaults(&required)
+	ensureValidatingWebhookConfigurationCaBundle(&required, *existing)
```

Comment:
nit: let's keep the `existing, required` ordering in the parameters here, for consistency. it's error-prone when two arguments with identical type have different ordering in similar-purpose methods

also I'd prefer the shorter `existing, required *admissionregv1.ValidatingWebhookConfiguration` instead of repeating the full type everywhere

Reply:
> also I'd prefer the shorter existing, required *admissionregv1.ValidatingWebhookConfiguration instead of repeating the full type everywhere

Thanks for the tip!

---

> let's keep the existing, required ordering in the parameters here, for consistency. it's error-prone when two arguments with identical type have different ordering in similar-purpose methods

The  two arguments of the function **don't** have identical types because one is a pointer `*admissionregv1.ValidatingWebhookConfiguration`, and the second is a structure `admissionregv1.ValidatingWebhookConfiguration`.

I suppose the name of the function `ensureValidatingWebhookConfigurationCaBundle` causes the problem.

It contains the word `ensure` which is used in a little bit different context than the other `ensure<Resource>` functions that contain the argument `modified *bool` to signal if the `existing` has been modified.

Here, the CVO would set the fields of `required` so as to not trigger an unnecessary update. It's more similar to `ensure<Resource>Defaults` functions but here we need the information from `existing` (the CA bundle) thus I put it as a second argument (I guess I was prioritizing the importance of arguments in the ordering).

But I understand the problem with the consistency and the error-prone use. Do you think swapping the parameters, as you mentioned, would be enough? Now that I am thinking about it, the current name of the function could cause confusion like "_It's ensuring that the CABundle fields of webhooks are the same like the other functions do_"

Reply:
> The two arguments of the function don't have identical types because one is a pointer 

doh, right, I overlooked that. That makes sense, because one is to be modified and one isn't. And the one being modified is always first, so it is consistent enough. The confusing part is that sometimes we modify `existing` and sometimes we modify `required`. The latter is defaulting. I agree that the names contribute to confusion, but they are not *obviously* wrong, so I'd perhaps just proceed.

If you feel like trying to improve it, then I  feel that doing all defaulting inside `ensureValidatingWebhookConfigurationDefaults` could be a step in a good direction. That method also iterates `required.Webhooks` so maybe it could do the `caBundle` determination, something like:

```go
func ensureValidatingWebhookConfigurationDefaults(required *admissionregv1.VWC, existing VWC) {
        var caBundle string
        if val, ok := <get annotation>; ok && val = "true" && len(existing.Webhooks) == 0 {
		caBundle = existing.Webhooks[0].ClientConfig.CABundle
	}
	for i := range required.Webhooks {
		ensureValidatingWebhookDefaults(&required.Webhooks[i])
		if caBundle != "" {
			required.Webhooks[i].ClientConfig.CABundle = caBundle
                }
	}
}
```

/shrug
But it's not that much better so I guess we may as well proceed with the PR. I'll /hold this if you want to explore some options, but feel free to unhold if you don't want to bother.

/hold
/lgtm
/approve

Reply:
That's also a possibility that ensures the function wouldn't be misused. But then the function name would be semantically incorrect since setting the CA bundles is not technically defaulting. I guess the best option would be to introduce new functions like `initiliazeRequired<Resource>` that would set the default values and would set things like these for the CVO to ignore. But that's reformatting on another level that should ideally be done for all the resources, and it's not that necessary.

I'll keep the PR on `hold` for some time to see if anybody has any suggestions or opinions.

Any suggestions are welcome! :heart: 

If there are no suggestions after ~24 hours, I'll `unhold` the PR.

Reply:
> should ideally be done for all the resources, and it's not that necessary.

Yeah that was my thinking as well: there are refactoring opportunities that would streamline the code and eliminate these issues but that's likely a separate effort and definitely not a must for a simple hotlooping bugfix.